src/safe.d

```diff
@@ -0,0 +1,175 @@
+/**
+ * Compiler implementation of the
+ * $(LINK2 http://www.dlang.org, D programming language).
+ *
+ * Copyright:   Copyright (c) 1999-2016 by Digital Mars, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(DMDSRC _safe.d)
+ */
+
+module ddmd.safe;
+
+import core.stdc.stdio;
+
+import ddmd.aggregate;
+import ddmd.dclass;
+import ddmd.declaration;
+import ddmd.dscope;
+import ddmd.expression;
+import ddmd.mtype;
+import ddmd.target;
+import ddmd.tokens;
+
+
+/*************************************************************
+ * Check for unsafe access in @safe code:
+ * 1. read overlapped pointers
+ * 2. write misaligned pointers
+ * 3. write overlapped storage classes
+ * Print error if unsafe.
+ * Params:
+ *      sc = scope
+ *      e = expression to check
+ *      readonly = if access is read-only
+ *      printmsg = print error message if true
+ * Returns:
+ *      true if error
+ */
+
+bool checkUnsafeAccess(Scope* sc, Expression e, bool readonly, bool printmsg)
+{
+    if (e.op != TOKdotvar)
+        return false;
+    DotVarExp dve = cast(DotVarExp)e;
+    if (VarDeclaration v = dve.var.isVarDeclaration())
+    {
+        if (sc.intypeof || !sc.func || !sc.func.isSafeBypassingInference())
+            return false;
+
+        auto ad = v.toParent2().isAggregateDeclaration();
+        if (!ad)
+            return false;
```

Comment:
Yeah, but this is a refactoring PR. That line was not covered before, either. Should not mix refactoring with semantic changes.